src/python/pants/backend/graph_info/tasks/cloc.py

```diff
@@ -32,59 +32,68 @@ def register_options(cls, register):
     register('--ignored', type=bool, fingerprint=True,
              help='Show information about files ignored by cloc.')
 
-  def _get_cloc_script(self):
-    return ClocBinary.global_instance().select(self.context)
-
   def console_output(self, targets):
     if not self.get_options().transitive:
       targets = self.context.target_roots
 
-    buildroot = get_buildroot()
+    # TODO: Work out a nice library-like utility for writing an argfile, as this will be common.
     with temporary_dir() as tmpdir:
-      # Write the paths of all files we want cloc to process to the so-called 'list file'.
-      # TODO: 1) list_file, report_file and ignored_file should be relative files within the
-      # execution "chroot", 2) list_file should be part of an input files Snapshot, and
-      # 3) report_file and ignored_file should be part of an output files Snapshot, when we have
-      # that capability.
-      list_file = os.path.join(tmpdir, 'list_file')
+      list_file = os.path.join(tmpdir, 'input_files_list')
+      input_files = set()
       with open(list_file, 'w') as list_file_out:
         for target in targets:
           for source in target.sources_relative_to_buildroot():
-            list_file_out.write(os.path.join(buildroot, source))
+            input_files.add(source)
+            list_file_out.write(source)
             list_file_out.write(b'\n')
+      list_file_snapshot = self.context._scheduler.capture_snapshots((
+        PathGlobsAndRoot(
+          PathGlobs(('input_files_list',), ()),
+          str(tmpdir),
+        ),
+      ))[0]
+
+    cloc_path, cloc_snapshot = ClocBinary.global_instance().hackily_snapshot(self.context)
+
+    # TODO: This should use an input file snapshot which should be provided on the Target object,
+    # rather than hackily re-snapshotting each of the input files.
+    # See https://github.com/pantsbuild/pants/issues/5762
+    input_pathglobs = PathGlobs(tuple(input_files), ())
+    input_snapshot = self.context._scheduler.product_request(Snapshot, [input_pathglobs])[0]
+
+    directory_digest = self.context._scheduler.merge_directories((
```

Comment:
Could refer to https://github.com/pantsbuild/pants/issues/5502 here and mention moving more of this into `@rules`.